### ISSUE TITLE
New Exceptional Error Handling Leads To Problems When Saving Note

### ISSUE THREAD

**Issue by lucidBrot** [GitJournal/GitJournal#941]
edit: sorry about the verbosity, I am typing as I observe. It turns out that half of what I wrote is not really relevant.

---

I am currently working on rebasing/recreating a PR on the latest master and because of this tried to run the commit `7fd1c9928773f5dcd7c1e042833957f93e83e187` in my android emulator.

### Steps to Reproduce
1. Install app from latest master commit in android emulator. (Built with `flutter run --flavor dev --debug` on a Pixel 2 API 29)
2. open an existing note or create a new one
3. modify
4. hit the save button
5. observe the error popup and console logs.

### Workaround
This issue can be avoided by enabling the "dart-only git implementation" option in the experimental settings.

### Error Info
![image](https://github.com/GitJournal/GitJournal/assets/4920691/47bac104-0951-45a8-b63d-a9f40deaaa7a)

```
I/flutter (20904): D NoteEditorState._saveNote: Note modified - saving
I/flutter (20904): D NoteEditorState._saveNote: Rebuilding Note's FileName
I/flutter (20904): D logEvent: Event.NoteAdded
I/flutter (20904): D GitJournalRepo.addNote.<ac>: Got addNote lock
I/GitJournalAndroid(20904): Called method gitAdd
I/GitJournalAndroid(20904): .  folderPath: /storage/emulated/0/journal-46489/
I/GitJournalAndroid(20904): .  filePattern: .
E/GitJournalLib(20904): gj_error_info: -1
E/GitJournalLib(20904): Error -1/10: invalid data in index - invalid entry
I/flutter (20904): E logException: Got Exception GitException: invalid data in index - invalid entrypackage:git_bindings/git_bindings.dart 67:7         GitRepo.add
I/flutter (20904): package:gitjournal/core/git_repo.dart 47:7          GitNoteRepository._add
I/flutter (20904): package:gitjournal/core/git_repo.dart 79:5          GitNoteRepository._addAllAndCommit
I/flutter (20904): package:gitjournal/repository.dart 556:7            GitJournalRepo.addNote.<fn>
I/flutter (20904): package:synchronized/src/basic_lock.dart 33:16      BasicLock.synchronized
I/flutter (20904): package:gitjournal/repository.dart 553:5            GitJournalRepo.addNote
I/flutter (20904): package:gitjournal/editors/note_editor.dart 538:9   NoteEditorState._saveNote
I/flutter (20904): package:gitjournal/editors/note_editor.dart 358:17  NoteEditorState._exitEditorSelected
I/flutter (20904):
D/InputConnectionAdaptor(20904): The input method toggled cursor monitoring off
```

### Solution
Not sure if the git bindings that do not use dart-git need to be supported. But if so, perhaps catching and ignoring the exceptions would be the right thing to do, since that seems to have been the old behaviour?

### My ToDos
- [ ] double-check on a clean emulator with no previous app data to ensure the issue is unrelated to any settings I might have set.

**Comment by lucidBrot** [comment]
Huh, alright excuse me: This is not as bad as it initially seemed. Clean installs with a new repo still work.

I can **not** reproduce the error on a clean emulator unless I `adb push` my test directory (which is not randomly generated data but my actual notes, so I can't just provide that as an example test case). When I do use my test directory however, I can reproduce the error with the default app settings except for "Store repo externally".

I will follow up once I've tried it on a rooted emulator so I can say what happens if the repo is stored internally.

**Comment by lucidBrot** [comment]
It is reproducible on a clean emulator with no custom app settings:

* initial creation of a note is fine.
* I then use `adb root; adb push mydir ` /data/user/0/io.gitjournal.gitjournal.dev/app_flutter/journal-55933/` to add my notes in the app's internal directory.
* On the next refresh of the app, caching clearly thinks still that there is only 1 file.
  ```
I/flutter ( 7310): D GitJournalRepo.syncNotes: Not syncing because RemoteRepo not configured
I/flutter ( 7310): D FileStorageCache.save: Saving MTimeCache: 1 items
I/flutter ( 7310): D FileStorageCache.save: Saving CTimeCache: 1 items
I/flutter ( 7310): E FileStorage.load: Failed to build path: priv-notes.bak/ctf.md nullnull
I/flutter ( 7310): E NotesFolderFS._load: NotesFolderFS FileStorage Failure Instance of 'FileStorageCacheIncomplete'package:gitjournal/core/file/file_storage.dart 63:7         FileStorage.load
I/flutter ( 7310): package:gitjournal/core/folder/notes_folder_fs.dart 296:34  NotesFolderFS._load
I/flutter ( 7310): package:synchronized/src/basic_lock.dart 33:16              BasicLock.synchronized
I/flutter ( 7310): package:gitjournal/core/folder/notes_folder_fs.dart 242:5   NotesFolderFS.loadRecursively
I/flutter ( 7310): dart:async                                                  Future.wait.<fn>
I/flutter ( 7310): package:gitjournal/core/folder/notes_folder_fs.dart 251:5   NotesFolderFS.loadRecursively
I/flutter ( 7310): package:gitjournal/repository.dart 278:9                    GitJournalRepo._loadNotes.<fn>
I/flutter ( 7310): package:synchronized/src/basic_lock.dart 33:16              BasicLock.synchronized
I/flutter ( 7310): package:gitjournal/repository.dart 338:7                    GitJournalRepo.syncNotes
I/flutter ( 7310): package:gitjournal/folder_views/folder_view.dart 577:5      syncRepo
I/flutter ( 7310):
I/AssistStructure( 7310): Flattened final assist data: 532 bytes, containing 1 windows, 3 views
  ```

**Comment by lucidBrot** [comment]
It seems that this problem was caused by having a nested `.git` directory somewhere in a subdirectory.

The old behaviour was to ignore such an exception. The new behaviour is to fail on saving with a warning that the data might have not been saved (but it has been saved, just not committed).

I don't think handling this is responsibility of GitJournal. It's just a weird edge-case I stumbled into. 

Switching the experimental setting "dart-only git implementation" to enabled *does* successfully work around this issue in the test case I had.
